lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/security/FileBasedSystemAccessControl.java

```diff
@@ -811,20 +824,32 @@ public void checkCanSetMaterializedViewProperties(SystemSecurityContext context,
     @Override
     public void checkCanGrantExecuteFunctionPrivilege(SystemSecurityContext context, String functionName, TrinoPrincipal grantee, boolean grantOption)
     {
+        Identity identity = context.getIdentity();
+        boolean allowed = functionRules.stream()
+                .filter(functionRule -> functionRule.matches(identity.getUser(), identity.getEnabledRoles(), identity.getGroups(), functionName))
+                .map(CatalogFunctionAccessControlRule::canGrantExecuteFunction)
+                .findFirst()
+                .orElse(false);
+        if (!allowed) {
+            String granteeAsString = format("%s '%s'", grantee.getType().name().toLowerCase(ENGLISH), grantee.getName());
+            denyGrantExecuteFunctionPrivilege(functionName, context.getIdentity(), granteeAsString);
+        }
     }
 
     @Override
     public void checkCanGrantExecuteFunctionPrivilege(SystemSecurityContext context, FunctionKind functionKind, CatalogSchemaRoutineName functionName, TrinoPrincipal grantee, boolean grantOption)
     {
-        switch (functionKind) {
-            case SCALAR, AGGREGATE, WINDOW:
-                return;
-            case TABLE:
-                // TODO (https://github.com/trinodb/trino/issues/12833) implement
-                String granteeAsString = format("%s '%s'", grantee.getType().name().toLowerCase(ENGLISH), grantee.getName());
-                denyGrantExecuteFunctionPrivilege(functionName.toString(), context.getIdentity(), granteeAsString);
+        AccessMode requiredCatalogAccess = functionKind == TABLE ? ALL : READ_ONLY;
```

Comment:
So what was a decision about appropriate mode?

Reply:
Right now we require `ALL` for `PTFs`, later we could improve that